### PR TITLE
Improved hint placement

### PR DESCRIPTION
**Pull request by SpaghettiToastBook**
Needs testing, but I thought I'd make a PR anyway.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 -   *Major* - Added: Title screen now shows a three-word representation of the seed hash.
 
+-   *Major" - Changed: The hint placement algorithm was changed so that hints are more likely to be useful.
```
Comment (henriquegemignani) [inline CHANGELOG.md]: There's a `"` instead of `*` here.

File: randovania/generator/filler/retcon.py [hunk 2]
```diff
@@ -85,6 +85,41 @@ def _resources_in_pickup(pickup: PickupEntry, current_resources: CurrentResource
...
+
+    patches = final_state.patches
+
+    state_sequence = [final_state]
+    while state_sequence[-1].previous_state:
+        state_sequence.append(state_sequence[-1].previous_state)
+    state_sequence = tuple(reversed(state_sequence))
+
+    indices_with_hints = {index for index in final_state.patches.pickup_assignment if
+                          _should_have_hint(final_state.patches.pickup_assignment[index].item_category)}
```
Comment (henriquegemignani) [inline randovania/generator/filler/retcon.py]: Can be simplified as:
```python
indices_with_hints = {index for index, pickup in final_state.patches.pickup_assignment.items()
                      if _should_have_hint(pickup.item_category)}
```

File: randovania/generator/filler/retcon.py [hunk 3]
```diff
@@ -85,6 +85,41 @@ def _resources_in_pickup(pickup: PickupEntry, current_resources: CurrentResource
...
+
+    sequence = []
+    for state in state_sequence[1:]:
+        new_indices = list((set(state.collected_pickup_indices) & indices_with_hints)
+                           - (set(state.previous_state.collected_pickup_indices) | set(patches.starting_items)))
```
Comment (henriquegemignani) [inline randovania/generator/filler/retcon.py]: `patches.starting_items` should only contains resources of the type `ITEM`. I don't think you need to use it here.

File: randovania/generator/filler/retcon.py [hunk 4]
```diff
@@ -85,6 +85,41 @@ def _resources_in_pickup(pickup: PickupEntry, current_resources: CurrentResource
...
+
+        new_logbook_assets = list(set(state.collected_scan_assets)
+                                  - (set(state.previous_state.collected_scan_assets) | set(patches.hints)))
+        rng.shuffle(new_logbook_assets)
+
+        sequence.append((new_indices, new_logbook_assets))
+
+    hints_placed = 0
+    index_list = sum((indices for indices, _ in sequence), [])
+    for indices, logbooks in sequence:
```
Comment (henriquegemignani) [inline randovania/generator/filler/retcon.py]: A comment explaining what each this section does should be helpful.

File: randovania/generator/filler/retcon.py [hunk 5]
```diff
@@ -157,16 +194,7 @@ def action_report(message: str):
...
-
                 next_state = reach.state.assign_pickup_to_index(action, pickup_index)
-                if current_uncollected.logbooks and _should_have_hint(action.item_category):
-                    hint_location: Optional[LogbookAsset] = rng.choice(list(current_uncollected.logbooks))
-                    next_state.patches = next_state.patches.assign_hint(hint_location,
-                                                                        Hint(HintType.LOCATION, None, pickup_index))
-                else:
-                    hint_location = None
-
-                print_retcon_place_pickup(action, game, pickup_index, hint_location)
```
Comment (henriquegemignani) [inline randovania/generator/filler/retcon.py]: This debug print is still relevant! :)

File: randovania/generator/filler/runner.py [hunk 6]
```diff
@@ -113,13 +114,12 @@ def fill_unassigned_hints(patches: GamePatches,
...
 
     new_hints = copy.copy(patches.hints)
-    possible_indices = list(patches.pickup_assignment.keys())
 
     for node in world_list.all_nodes:
         if isinstance(node, LogbookNode):
             logbook_asset = node.resource()
             if logbook_asset not in new_hints:
-                new_hints[logbook_asset] = Hint(HintType.LOCATION, None, rng.choice(possible_indices))
+                new_hints[logbook_asset] = Hint(HintType.LOCATION, PrecisionPair.joke(), PickupIndex(-1))
```
Comment (henriquegemignani) [inline randovania/generator/filler/runner.py]: I think this will make Minimal Checking almost not have hints!

**Comment by SpaghettiToastBook** [comment]
I'm pretty happy with the hint placement algorithm now, but these things still need to be done:
- [ ] Write tests for `place_hints`
- [ ] Have game patches serializer tests not fail
- [ ] Maybe make the hints a bit less good, if they're too good

File: randovania/generator/filler/hint_placement.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    state_sequence = [final_state]
+    while state_sequence[-1].previous_state:
+        state_sequence.append(state_sequence[-1].previous_state)
+    state_sequence = tuple(reversed(state_sequence))
+
+    possible_hint_indices = {index for index, pickup in patches.pickup_assignment.items()
+                             if pickup.can_get_hint or index in _SPECIAL_HINTS}
+
+    sequence: List[Union[PickupIndex, LogbookAsset]] = []
+    for state in state_sequence[1:]:
```
Comment (henriquegemignani) [inline randovania/generator/filler/hint_placement.py]: This for loop could be refactored out into another function.

Comment (SpaghettiToastBook) [reply randovania/generator/filler/hint_placement.py]: ✓

**Pull request closed without merging**